jax/interpreters/pxla.py

```diff
@@ -754,6 +755,9 @@ def dynamic_fun(dummy, *args):
   # get_default_device_assignment() returns 2D assignment, caller may have
   # provided 1D list of devices).
   device_assignment = tree_map(lambda d: d.id, devices)
+  # Convert to 2D in case it's 1D and we have > 1 partitions.
+  device_assignment = onp.array(device_assignment).reshape(
```

Comment:
At least in the nested pmap case, device_assignment order seems to matter a lot for all-reduce times.
This could be true for sharded_jit as well - worth keeping an eye out for this.

Reply:
Agreed. I'd like to create a fancier "default device assignment" API that we can implement in XLA, but it's kind of hard to describe the constraints (something like, these replicas need to be on these hosts).

Reply:
I'll send you the code I'm currently using for generating device_assignment ordering (and the too-loose constraints it's using), in case it's helpful.

Reply:
Thanks, I think it will be!